drivers: gpio: Introduce SC18IS606 GPIO Extension drivers

zacck opened the pull request:
Introduce I2C to SPI bridge driver GPIO  support for the SC18IS6906 bridge from NXP.

This adds an GPIO controller available on the bridge from NXP. Ideally this enables one to add an GPIO extension through i2c

zacck commented on drivers/gpio/gpio_sc18is606.c:
```diff
@@ -0,0 +1,252 @@
+	ret = nxp_sc18is606_transfer(cfg->bridge, buf, sizeof(buf), NULL, 0, NULL);
+	if (ret < 0) {
+		LOG_ERR("Failed to configure GPIO (%d)", ret);
+	}
```
@kartben  so is this

JarmouniA commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -144,3 +144,8 @@ if(CONFIG_GPIO_SC18IM704)
 endif()
+
+if(CONFIG_GPIO_SC18IS606)
+  zephyr_library_include_directories(${ZEPHYR_BASE}/drivers)
```
```suggestion
```

JarmouniA commented on drivers/gpio/Kconfig.sc18is606:
```diff
@@ -0,0 +1,19 @@
+config GPIO_SC18IS606
+	bool "NXP SC18IS606 GPIO Controller driver"
+	default y
+	depends on SPI_SC18IS606
```
Perhaps you meant
```
	depends on SC18IS606
```

zacck replied:
It does need the header from the SPI device... the mutex and the transactions are done there.

JarmouniA replied:
> Unless I missed a way to actually have a common file for mfd devices that belongs to base and just the base

What do you mean?

mmahadevan108 replied:
`drivers/mfd` might be a good place for the common driver to communicate. Currently you have an API defined inside the SPI driver.

JarmouniA commented on dts/bindings/mfd/nxp,sc18is606.yaml:
```diff
@@ -24,6 +24,14 @@ description: |
         #size-cells = <0>;
       };
+
+      sc18is606_gpio: sc18is606_gpio {
```
```suggestion
      sc18is606_gpio: gpio {
```
https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#generic-names-recommendation

mmahadevan108 commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -102,6 +102,7 @@ zephyr_library_sources_ifdef(CONFIG_GPIO_SAM gpio_sam.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_SAM0 gpio_sam0.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_SAM4L gpio_sam4l.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_SAM_PIO4 gpio_sam_pio4.c)
+zephyr_library_sources_ifdef(CONFIG_GPIO_SC18IS606 gpio_sc18is606.c)
```
You probably need to look into why you have changes to GPIO in this commit. All GPIO related changes should be in the later commit related to GPIO.

zacck replied:
@this change corresponds with stopping the dependency on the spi_header, I can try and amend it so it is in the other commit.

JarmouniA commented on include/zephyr/drivers/mfd/sc18is606.h:
```diff
@@ -4,8 +4,8 @@
 
-#ifndef ZEPHYR_DRIVERS_I2C_SPI_SC18IS606_H_
-#define ZEPHYR_DRIVERS_I2C_SPI_SC18IS606_H_
+#ifndef ZEPHYR_DRIVERS_SC18IS606_H_
```
```suggestion
#ifndef INCLUDE_ZEPHYR_DRIVERS_MFD_SC18IS606_H_
```

JarmouniA commented on dts/bindings/gpio/nxp,sc18is606-gpio.yaml:
```diff
@@ -0,0 +1,24 @@
+# SPDX-License-Identifier: Apache-2.0
+
+
+description: GPIO Controller part for the SC18IS606 bridge.
```
```suggestion
description: GPIO Controller part for the SC18IS606 bridge.
             Any compatible node shall be a child of a "nxp,sc18is606" compatible node.
```

JarmouniA commented on drivers/gpio/Kconfig.sc18is606:
```diff
@@ -0,0 +1,19 @@
+	help
+	   SC18IS606 GPIO Controller Init priority
+
+	   Note: Has to be greater than  the parent SC18IS606 bridge init priority
```
```suggestion
	   Note: Has to be greater than the parent SC18IS606 bridge init priority
```

JarmouniA commented on drivers/gpio/gpio_sc18is606.c:
```diff
@@ -12,24 +12,26 @@
-#define SC18IS606_GPIO_WRITE	0xF4
-#define SC18IS606_GPIO_READ	0xF5
-#define SC18IS606_GPIO_ENABLE	0xF6
-#define SC18IS606_GPIO_CONF	0xF7
```
No unnecessary formatting changes please, esp. not mixed with functional ones in the same commit.

JarmouniA reviewed (changes requested):
The commit decoupling common parts from SPI driver should come first, then GPIO driver can be added.

zacck committed:
drivers: mfd: Add base support for NXP SC18IS606

Add support for base sc18is606  device and move dependencies to
mfd header rather than the SPI device.

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
drivers: gpio: Introduce SC18IS606 GPIO Controller

Added a driver implementation for the GPIO Controller
on the sc18is606

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
tests: gpio: Add overlay for sc18is606 gpio

Apply device tree overlay for SC18IS606 mfd tests
in build_all

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

JarmouniA commented on drivers/mfd/mfd_sc18is606.h:
```diff
@@ -0,0 +1,109 @@
+/**
+ * @file
+ * @brief MFD SC18IS606 interface
+ * @ingroup mfd_interfaces
```
```suggestion
```

JarmouniA commented on dts/bindings/mfd/nxp,sc18is606.yaml:
```diff
@@ -24,6 +24,14 @@ description: |
         #size-cells = <0>;
       };
+
+      gpio_ext: gpio {
```
Shouldn't be in the 1st commit

JarmouniA commented on drivers/spi/Kconfig.sc18is606:
```diff
@@ -4,7 +4,7 @@
 	bool  "NXP SC18IS606 SPI controller driver"
 	default y
-	select I2C
+	depends on MFD_SC18IS606
```
```suggestion
        select MFD
	depends on MFD_SC18IS606
```

zacck replied:
I will try this again but I think it breaks because MFD is already mentioned might be a different variant

JarmouniA replied:
> because MFD is already mentioned

Not sure what you mean

JarmouniA commented on drivers/spi/spi_sc18is606.c:
```diff
@@ -16,8 +16,10 @@ LOG_MODULE_REGISTER(spi_sc18is606, CONFIG_SPI_LOG_LEVEL);
 #include <zephyr/drivers/gpio.h>
 #include <zephyr/drivers/i2c.h>
 #include <zephyr/drivers/spi.h>
+
```
```suggestion
```
